refactor(options)!: unify interfaces for setting options

famiu opened the pull request:
Problem: There is a separate function for setting an option for each type, which are currently `set_string_option`, `set_num_option`, `set_bool_option`, `do_set_option_value`, `do_set_num_option`, `do_set_bool_option`, `do_set_option_string`. This leads to a lot of unnecessary code duplication and also vastly reduces maintainability and extensibility.

Solution: Unify most ways of setting options to a single `set_option` function, cleaning up a lot of code duplication and greatly simplifying the options code, thus making it a lot more extensible and making it easier to add more option types. This PR unifies everything besides some functions used for string options (`set_string_option`, `did_set_string_option`, etc.). Unification of the code paths for setting string options and all other options will be done in a follow-up PR.

BREAKING CHANGE: This breaks the `OptionSet` autocommand, as the `v:` values associated with it (`v:option_new`, `v:option_old`, `v:option_oldlocal` and `v:option_oldglobal`) are now the same type as the option, instead of all option values being converted to strings.

famiu committed:
refactor(options): unify set_num_option and set_bool_option

bfredl commented on src/nvim/option.c:
```diff
@@ -3733,67 +3681,136 @@ int get_option_value_strict(char *name, int64_t *numval, char **stringval, int o
+    optset_T did_set_cb_args = (value.type == kOptValTypeNumber)
+                                 ? (optset_T){ .os_varp = varp,
+                                               .os_flags = opt_flags,
+                                               .os_oldval.number = old_value.data.number,
+                                               .os_newval.number = value.data.number,
```
I would rather have two ternaries, inside `.os_oldval` and .os_newval` respectively, so the rest of the code can be shared.

famiu replied:
Ah, that would indeed be a much smarter way of doing it

famiu committed:
refactor(options)!: make OptionSet `v:` values use typval

BREAKING CHANGE: This breaks the OptionSet autocommand, as the `v:` values associated with it (`v:option_new`, `v:option_old`, `v:option_oldlocal` and `v:option_oldglobal`) are now the same type as the option, instead of all option values being converted to strings.

famiu committed:
refactor(options): remove `os_op`

famiu committed:
refactor(options): deduplicate `do_set_option_string`

Reduce code duplication between `do_set_option_string` and `set_string_option` by making the former call the latter within itself.

famiu committed:
refactor(options): deduplicate the `do_set_option` codepath

famiu committed:
refactor(options): rename `empty_option` to `empty_string_option`

famiu committed:
refactor(options): make `os_oldval` and `os_newval` use `OptValData`

MariaSolOs commented on src/nvim/eval.c:
```diff
@@ -8694,7 +8704,7 @@ char *do_string_sub(char *str, char *pat, char *sub, typval_T *expr, const char
   char *ret = xstrdup(ga.ga_data == NULL ? str : ga.ga_data);
   ga_clear(&ga);
-  if (p_cpo == empty_option) {
+  if (p_cpo == empty_string_option) {
```
Shouldn't you use `strcmp` here?

famiu replied:
Nope, we're comparing the literal addresses here since `empty_string_option` often isn't copied

famiu committed:
refactor(options): make `did_set_shada` a normal option callback

zeertzjq commented on runtime/doc/news.txt:
```diff
@@ -71,6 +71,9 @@ The following changes may require adaptations in user config or plugins.
   defined by LSP, and hence previously parsed snippets might now be considered
   invalid input.
 
+• |v:option_new|, |v:option_old|, |v:option_oldlocal|, |v:option_oldglobal|
+  now have the type of the option instead of always being strings.
+
```
Also mention in `vim_diff.txt`?

zeertzjq commented on src/nvim/option.c:
```diff
@@ -1486,8 +1402,21 @@ static void do_set_option(int opt_flags, char **argp, bool *did_show, char *errb
+  }
+
+  *errmsg = set_option(opt_idx, varp, newval, opt_flags, op == OP_NONE, errbuf, errbuflen);
+  // `set_option` copies the new option value, so it needs to be freed  here.
```
```suggestion
  // `set_option` copies the new option value, so it needs to be freed here.
```

zeertzjq commented on src/nvim/option.c:
```diff
@@ -1258,34 +1134,144 @@ static int validate_opt_idx(win_T *win, int opt_idx, int opt_flags, uint32_t fla
+    if (nextchar == '!') {
+      switch (oldval.data.boolean) {
+      case kNone:
+        abort();
```
This triggers the abort:
```vim
setlocal autoread<
setlocal autoread!
```

zeertzjq replied:
```suggestion
        newval_bool = kNone;
        break;
```

zeertzjq commented on src/nvim/option.c:
```diff
@@ -1328,34 +1134,144 @@ static int validate_opt_idx(win_T *win, int opt_idx, int opt_flags, uint32_t fla
+    // other        error
+    arg++;
+    if (nextchar == '&') {
+      newval_num = (varnumber_T)options[opt_idx].def_val;
```
```suggestion
      newval_num = (OptInt)(intptr_t)options[opt_idx].def_val;
```

zeertzjq commented:
Otherwise LGTM

justinmk commented on runtime/doc/news.txt:
```diff
@@ -71,6 +71,9 @@ The following changes may require adaptations in user config or plugins.
   defined by LSP, and hence previously parsed snippets might now be considered
   invalid input.
 
+• |v:option_new|, |v:option_old|, |v:option_oldlocal|, |v:option_oldglobal|
```
Let's add some extra context 

```suggestion
• OptionSet autocommand args (|v:option_new|, |v:option_old|, |v:option_oldlocal|, |v:option_oldglobal|
```

justinmk commented on src/nvim/eval/vars.c:
```diff
@@ -1897,6 +1897,45 @@ static OptVal tv_to_optval(typval_T *tv, const char *option, uint32_t flags, boo
   return value;
 }
 
+/// Consume an option value and convert it to typval.
```
"consume" implies that some state is being advanced (e.g. in a stream/queue/etc), is that the case here?

famiu replied:
Well, in this case, it just means it doesn't copy the value and instead uses the memory allocated by the OptVal.

justinmk replied:
Sounds more like "read"? Should avoid words that have extra meaning unless the extra meaning is relevant.

justinmk commented on src/nvim/globals.h:
```diff
@@ -789,10 +789,9 @@ EXTERN char *escape_chars INIT(= " \t\\\"|");  // need backslash in cmd line
-EXTERN char *empty_option INIT(= "");
+// When a string option is NULL (which only happens in out-of-memory situations), it is set to
+// empty_string_option, to avoid having to check for NULL everywhere.
+EXTERN char *empty_string_option INIT(= "");
```
Potential future improvement: is this even possible in Nvim? Is the OOM being referred-to happening during startup, or could it happen after startup? Does it even matter?

famiu replied:
I'm not sure, I would have to look into it. Either way, we could add just ref-counted Strings which removes the need for this

justinmk commented on src/nvim/option.c:
```diff
@@ -1328,34 +1134,145 @@ static int validate_opt_idx(win_T *win, int opt_idx, int opt_flags, uint32_t fla
 
-static void do_set_option(int opt_flags, char **argp, bool *did_show, char *errbuf,
-                          size_t errbuflen, const char **errmsg)
+static void do_set_one_option(int opt_flags, char **argp, bool *did_show, char *errbuf,
```
"one options" as opposed to what? why rename this?

famiu replied:
It's mostly meant to be analogous to `do_cmdline` and `do_one_cmd`, where the former operates on the entire cmdline and the latter only does one cmd at a time. `do_set` operates on the entire set statement whereas `do_set_one_option` operates on one option at a time.

clason replied:
So `do_one_[set_]option`?

justinmk replied:
Oh interesting. I think it's worth adding a one-line docstring to this function that explains the idea. Something like:

    /// Sets one option (as opposed to many options in one :set call, ":set a=b c=d ...").

famiu replied:
sure, that works too

famiu replied:
@justinmk So should I keep the name `do_set_one_option` or should I change it to `do_one_set_option` as clason suggested?

clason replied:
There's also `do_one_arg` in `arglist.c` -- two is coincidence, but three makes a pattern ;)

justinmk commented on src/nvim/option.c:
```diff
@@ -1486,8 +1403,21 @@ static void do_set_option(int opt_flags, char **argp, bool *did_show, char *errb
+    return;
+  }
+
+  OptVal newval = get_option_newval(opt_idx, opt_flags, prefix, argp, nextchar, op, flags, varp,
```
(off topic for this PR) AFAIR it's very unusual in the codebase to hide a pointer in a type. I.e. I would expect this to be:

```suggestion
  OptVal *newval = get_option_newval(opt_idx, opt_flags, prefix, argp, nextchar, op, flags, varp,
```

Is there a technical reason we didn't do this for `OptVal`?

famiu replied:
Well, it works similarly to the `Object` struct

justinmk commented on src/nvim/option.c:
```diff
@@ -3731,69 +3598,148 @@ int get_option_value_strict(char *name, int64_t *numval, char **stringval, int o
+                       optval_as_object(value_copy));
+  }
+
+  comp_col();                       // in case 'columns' or 'ls' changed
```
this kind of indentation only makes sense if it's aligned with other comments.
```suggestion
  comp_col();  // in case 'columns' or 'ls' changed
```

justinmk commented on src/nvim/option.c:
```diff
@@ -3731,69 +3598,148 @@ int get_option_value_strict(char *name, int64_t *numval, char **stringval, int o
+  // option setting the "local value" in fact sets the global value (since there is only one value).
+  //
+  // TODO(famiu): This needs to be changed to use the current type of the old value instead of
+  // value.type, when multi-type options are odded.
```
typo?

also is this TODO tracked somewhere, like in a tracking issue for the planned options-related work?

famiu replied:
Yup, that is a typo. And yes, it is not tracked. Should I make a tracking issue detailing the things I wish to work on?

justinmk replied:
> Should I make a tracking issue detailing the things I wish to work on?

That could be helpful, unless you have them in your head and will address them in a nearby PR :)

famiu replied:
I think it's best if I make an issue so I can have a concrete roadmap of things to work on, and others can know what I plan to do with this as well. Will do that in a bit

justinmk reviewed (approved):
LGTM generally, ~~but wondering if there is a tracking issue that tracks the TODOs~~ https://github.com/neovim/neovim/issues/25672
